### DISCUSSION TITLE
Touch suddenly became unavailable

### DISCUSSION THREAD

**Discussion by MikeyMoMo** [Bodmer/TFT_eSPI#1453] [Q&A - Touch]
I have been working on a touch application for 2 or 3 weeks.  My first, so I was learning and developing at the same time and had other duties, too.  A few days ago, after several weeks of playing with this code, I turned on the computer and, suddenly, it it telling me that touch is not available.  I have played all around with the libraries and even uninstalled and reinstalled Arduino IDE, current version.  I cleaned out the Documents\Arduino folder.  It cleaned out the Program Files\Arduino folder.  It is as if I have a totally fresh install of the environment.  It is for an ESP32 and always has been.  No change, there.

No change from one night shutting down to the next day starting up.  What happened?  Here are the error lines.  These come from from sample code: "...Documents\Arduino\libraries\TFT_eSPI\examples\480 x 320\Touch_Controller_Demo:
```
C:\Users\Mikey\Documents\Arduino\libraries\TFT_eSPI\examples\480 x 320\Touch_Controller_Demo\Touch_Controller_Demo.ino: In function 'void setup()':
Touch_Controller_Demo:43:9: error: 'class TFT_eSPI' has no member named 'setTouch'
     tft.setTouch(calibrationData);
         ^
Touch_Controller_Demo:46:9: error: 'class TFT_eSPI' has no member named 'calibrateTouch'
     tft.calibrateTouch(calibrationData, TFT_WHITE, TFT_RED, 15);
         ^
C:\Users\Mikey\Documents\Arduino\libraries\TFT_eSPI\examples\480 x 320\Touch_Controller_Demo\Touch_Controller_Demo.ino: In function 'void loop()':
Touch_Controller_Demo:63:11: error: 'class TFT_eSPI' has no member named 'getTouch'
   if (tft.getTouch(&x, &y)) {
           ^
Using library FS at version 1.0 in folder: C:\Users\Mikey\AppData\Local\Arduino15\packages\esp32\hardware\esp32\1.0.6\libraries\FS 
Using library SPI at version 1.0 in folder: C:\Users\Mikey\AppData\Local\Arduino15\packages\esp32\hardware\esp32\1.0.6\libraries\SPI 
Using library TFT_eSPI at version 2.3.70 in folder: C:\Users\Mikey\Documents\Arduino\libraries\TFT_eSPI 
Using library SPIFFS at version 1.0 in folder: C:\Users\Mikey\AppData\Local\Arduino15\packages\esp32\hardware\esp32\1.0.6\libraries\SPIFFS 
exit status 1
'class TFT_eSPI' has no member named 'setTouch'

```

**Comment by Bodmer** [comment]
Typically this error means that the touch chip select pin is not defined.

Run Read_User_Setup and check the compiler is picking up your setup.

If you load a new version of the TFT_eSPI library, then your setup may have been over-written.

**Reply by MikeyMoMo** [reply]
Somehow, TWICE NOW, a second TFT library has gotten installed.  I am using TFT_eSPI-master and a TFT_eSPI library is coming in from somewhere and that's the one the compiler has been selecting.  I have removed it twice, now.  After removing it, the properly configured library is selected and compiled and I am running again.  After I deleted it, I assumed it would stay gone and had not looked, again.  It was back and killing me some more.  It is gone again.  Now I know what to look for when compiling stops, again.

I would hope there could be a better error message when TOUCH_CS is not defined that would help the clueless user (me, sometimes) find the actual problem.  The message I was seeing did not lead me to the solution.

Thanks for the speedy answer.  I expect you are plenty busy and I appreciate it!!  I also hope you found my post about the setups for the ILI9486 that I am using.  It was supposed to be an ILI9341 but turned out not to be so I created a Setups file for it and added an automatic (controlled by #define) display inversion since it is needed on every program that uses this display.  Posted, elsewhere, about 10 days ago on this board.

I am running now.

Mike

**Reply by Bodmer** [reply]
There is a warning message here;
https://github.com/Bodmer/TFT_eSPI/blob/master/TFT_eSPI.h#L973

Not sure why it is not shown  unless warnings have been disabled...

**Comment by BrianB03** [comment]
Hi MikeyMoMo / Bodmer,
I have an identical fault! " .... 'class TFT_eSPI' has no member named 'getTouch'

My code has been working OK previously, now when I try to compile I get these same errors. (changed nothing otherwise)
Sample code "TFT_eSPI/examples/480x320/Keypad_480x320/Keypad_480x320.ino" will not run either? 

Stumped!
Brian

**Reply by Bodmer** [reply]
See previous answer.
https://github.com/Bodmer/TFT_eSPI/discussions/1453#discussioncomment-1711812

**Comment by MikeyMoMo** [comment]
I will look at the example.

My S3 dosplay has been destroyed by the latest update.

**Reply by elgerg** [reply]
Thats hardly a fair comment especially as this is well documented:
https://github.com/bodmer/TFT_eSPI/?tab=readme-ov-file#tips

**Comment by BrianB03** [comment]
Hi Phil,
My setup is Arduino,RP2040,TFT_eSPI and a 480x320 screen.
Still struggling .. here's a copy of error messages when running KeyPad_480x320:

In file included from /Users/brianbrks/Documents/Arduino/libraries/TFT_eSPI/TFT_eSPI.cpp:31:
/Users/brianbrks/Documents/Arduino/libraries/TFT_eSPI/Processors/TFT_eSPI_RP2040.c:18:48: error: 'TFT_MISO' was not declared in this scope; did you mean 'TFT_DISPON'?
   18 |     SPIClassRP2040 spi = SPIClassRP2040(SPI_X, TFT_MISO, -1, TFT_SCLK, TFT_MOSI);
      |                                                ^~~~~~~~
      |                                                TFT_DISPON
/Users/brianbrks/Documents/Arduino/libraries/TFT_eSPI/Processors/TFT_eSPI_RP2040.c:18:62: error: 'TFT_SCLK' was not declared in this scope
   18 |     SPIClassRP2040 spi = SPIClassRP2040(SPI_X, TFT_MISO, -1, TFT_SCLK, TFT_MOSI);
      |                                                              ^~~~~~~~
/Users/brianbrks/Documents/Arduino/libraries/TFT_eSPI/Processors/TFT_eSPI_RP2040.c:18:72: error: 'TFT_MOSI' was not declared in this scope
   18 |     SPIClassRP2040 spi = SPIClassRP2040(SPI_X, TFT_MISO, -1, TFT_SCLK, TFT_MOSI);
      |                                                                        ^~~~~~~~

exit status 1

Compilation error: exit status 1

**Reply by Bodmer** [reply]
If the display is used with a parallel interface then touch support will not be available, so a separate touch library will be needed.

**Comment by BrianB03** [comment]
Hi Phil,
Sorted at last!
Many thanks for your responses .. The attached file explains ..
Brian
[TFTLibsConfig.docx](https://github.com/Bodmer/TFT_eSPI/files/14851907/TFTLibsConfig.docx)
